lib/elixir/lib/kernel.ex

```diff
@@ -3210,6 +3210,11 @@ defmodule Kernel do
   """
   defmacro defstruct(fields) do
     quote bind_quoted: [fields: fields] do
+      if Module.get_attribute(__MODULE__, :struct) do
+        raise CompileError, file: __ENV__.file, line: __ENV__.line, description: "" <>
```

Comment:
You don't need to raise a compile error. A regular error (argument or runtime) would be fine because it will point to the exact line and stacktrace. CompileError are for things that happen later on based on some metadata.